hibernate-core/src/main/java/org/hibernate/type/format/StringJsonDocumentReader.java

```diff
@@ -0,0 +1,550 @@
+						default:
+							throw new IllegalStateException( "unexpected read ["+
+															this.jsonString.substring( this.jsonValueStart,this.jsonValueEnd )+
+															"] in current processing state " +
+															this.processingStates.getCurrent() );
+					}
+			}
+		}
+		// no way we get here.
+		return null;
+	}
+
+	/**
+	 * Gets the type of unquoted value.
+	 * We assume that the String value follows JSON specification. I.e unquoted value that starts with 't' can't be anything else
+	 * than <code>true</code>
+	 * @param jsonValueChar the value
+	 * @return the type of the value
+	 */
+	private JsonDocumentItemType getUnquotedValueType(char jsonValueChar) {
+		switch(jsonValueChar) {
+			case 't': {
+				//true
+				return JsonDocumentItemType.BOOLEAN_VALUE;
+			}
+			case 'f': {
+				//false
+				return JsonDocumentItemType.BOOLEAN_VALUE;
+			}
+			case 'n' : {
+					// null
+					return JsonDocumentItemType.NULL_VALUE;
+				}
+			case '-':
+			case '0':
+			case '1':
+			case '2':
+			case '3':
+			case '4':
+			case '5':
+			case '6':
+			case '7':
+			case '8':
+			case '9': {
+				return JsonDocumentItemType.NUMERIC_VALUE;
+			}
+			default :
+				return JsonDocumentItemType.VALUE;
+			}
+	}
+
+	private void moveBufferPosition(int shift) {
+		this.position += shift;
+	}
+
+	/**
+	 * Moves the current position to a given character, skipping any whitespace
+	 * We expect the character to be the next non-blank character in the sequence
+	 * @param character the character we should stop at.
+	 * @throws IllegalStateException if we encounter an unexpected character other than white spaces before the desired one.
+	 */
+
+	private void moveTo(char character) throws IllegalStateException {
+		int pointer = this.position;
+		while ( pointer < this.limit) {
+			char c = this.jsonString.charAt( pointer );
+			if ( c == character) {
+				this.position = pointer == this.position?this.position:pointer - 1;
+				return;
+			}
+			if (!Character.isWhitespace(c)) {
+				// we did find an unexpected character
+				// let the exception raise
+				//this.json.reset();
+				break;
+			}
+			pointer++;
+		}
+		throw new IllegalStateException("character [" + character + "] is not the next non-blank character");
+	}
+
+	/**
+	 * Goes through the JSON string to locate a character.
+	 * Escaped characters are taken into account.
+	 * ex: on 'AB\"C"' this method returns 5 (not 3)
+	 *
+	 * @param character character to be found
+	 * @return the position of the character or -1 if not found.
+	 */
+	private int locateCharacter(char character) {
+		int pointer = this.position;
+
+		while ( pointer < this.limit) {
+			final char c = this.jsonString.charAt( pointer );
+			if (c == ESCAPE_CHAR) {
+				// We encountered an escape character.
+				// We should just skip the next one as it is either the expected character
+				// but as escaped one, we should ignore it, either this is something else
+				// and we should ignore it also
+				pointer += 2;
```

Comment:
It seems this fails to account for unicode escape sequences. I assume that the following fragment will fail: " \u0020"

Reply:
Hi @beikov,
What do you have in mind?  I tried to process an object like 
```json
{
	"myUnicode1": "\u0074\u0068\u0069\u0073\u005f\u0069\u0073\u005f\u0075\u006e\u0069\u0063\u006f\u0064\u0065",
	"myUnicode2": "this_\u0069\u0073_unicode",
	"myUnicode3": "this_is_unicode"
}
```
And everything is fine.